var/spack/repos/builtin/packages/petsc/package.py

```diff
@@ -105,6 +105,44 @@ class Petsc(Package):
             description='Enable when mpiexec is not available to run binaries')
     variant('valgrind', default=False,
             description='Enable Valgrind Client Request mechanism')
+    variant('gmp', default=False,
+            description='Enable GMP support')
```

Comment:
We'll have to trim this list as some of these packages are primarily dependencies of dependencies.

For ex: zlib is a dependency of hdf5 - so its not listed as a variant - but as:
```
    depends_on('zlib', when='+hdf5')
```
In this case gmp is a dependency of mpfr  [which hardly gets tested].

BTW: Are these changes tested - i.e do spack builds work with these variants enabled?

Reply:
It is possible to have petsc build with gmp and not with mpfr, however, you are correct, building with mpfr requires gmp to be enabled. That is not reflected in the initial PR so would need to be corrected. I did not add a zlib variant for two reasons,
1. It is already the "default" because `+hdf5` is set by default.
1. It may be that zlib should be enabled by default and set as a dependency, but I had not explored that.

With regard to testing, our site's non-spack built petsc-3.10.3 was built including the following variants in this PR:
- netcdf
- mpfr
- libpng
- libjpeg
- gmp

My initial motivation for this PR was to add netcdf support. I then added the other things that I built previous versions of petsc with, and finally added others while I was in there. If any of those seem unneeded/unuseful/silly then I can pull them out. My main desire is for netcdf support, and to a lesser degree, parallel-netcdf.

Reply:
I think you've missed @balay's point. PETSc does not use GMP. PETSc's configure knows about it so it can find/build it as a dependency for MPFR (which is used in one rarely-used function).

Reply:
Note: the premise here his: we need variants only for packages that PETSc directly interfaces with.

The reference to hf5, zlib is an example of the case of a dependency of a dependency [as currently implemented in spack]

Also note: PETSc does have an interface to netcdf  parallel-netcdf. They are dependencies for other pacakges. (i.e a varaint should be added for them)

* pnetcdf is a dependency for exodusii, netcdf (optional)
* netcdf is a dependency for ADIOS (optional),moab (optional),Trilinos (alternate way of installing exodus,ML etc)

And generally - with shared libraires - one doesn't need to worry about dependencies of dependencies. But in static builds we do - hence this complexity in petsc configure. [apart from the fact that petsc configure can build these dependencies on its own]

Reply:
Well, petsc can build with and link gmp independently of mpfr. Whether it uses it is another story, but,
```
--with-gmp=1 --with-mpfr=0
```
will produce a petsc library linked with libgmp. In any event, if people think it is not a necessary variant, it can be removed.

Reply:
I would call this a bug in PETSc configure that isn't worth "fixing", except perhaps to convince people who want to turn "everything" on. It's an artifact of PETSc's configure being a sort of package manager. It's easy to check that Satish and I aren't lying to you about PETSc not using GMP. If you add this as a Spack variant, it would only serve to mislead more people.

Reply:
To illustrate:
```
balay@sb /home/balay/junk
$ cat sz.c
#include <stdio.h>
int main()
{
printf("test\n");
return 0;
}
balay@sb /home/balay/junk
$ gcc sz.c
balay@sb /home/balay/junk
$ ldd a.out
	linux-vdso.so.1 (0x00007ffcae5ef000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f683e203000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f683e3f5000)
balay@sb /home/balay/junk
$ mpicc sz.c
balay@sb /home/balay/junk
$ ldd a.out
	linux-vdso.so.1 (0x00007ffcfd7e8000)
	libmpi.so.12 => /home/balay/soft/mpich-3.3.2/lib/libmpi.so.12 (0x00007f9e199af000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f9e197bf000)
	libgfortran.so.5 => /lib64/libgfortran.so.5 (0x00007f9e194f7000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f9e193b1000)
	libOpenCL.so.1 => /lib64/libOpenCL.so.1 (0x00007f9e1938f000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f9e1936d000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f9e19360000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f9e19345000)
	libquadmath.so.0 => /lib64/libquadmath.so.0 (0x00007f9e192fb000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9e19ef2000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f9e192f4000)
```

Reply:
i.e just because `libmpi.so.12` is linked in does not imply the above test code is dependent on MPI

Reply:
I believe you, and always did, I was just presenting what I was observing. It sounds to me like gmp and mpfr variants can be nixed.

Reply:
```
$ git grep PETSC_HAVE_MPFR
src/dm/dt/interface/dt.c:#if defined(PETSC_HAVE_MPFR)
src/dm/dt/interface/dt.c:#if defined(PETSC_HAVE_MPFR)
src/dm/dt/tests/ex3.c:#if defined(PETSC_HAVE_MPFR)
```
gmp should be nixed. mpfr can be kep [however configure will insist on passing --with-gmp option for -with-mpfr to work]

netcdf, pnetcdf variants should be nixed aswell.

Will check the complete list...

Reply:
I just saw the most recent comment from @balay. I am somewhat confused by the following:
>Note: the premise here is: we need variants only for packages that PETSc directly interfaces with.
...
Also note: PETSc does have an interface to netcdf parallel-netcdf. They are dependencies for other pacakges. (i.e a varaint should not be added for them)

The first statement indicates that a variant would be viable if there is an interface but the second statement seems to contradict that. Or, did you mean that, "PETSc does **not** have an interface to netcdf parallel-netcdf"?

Sorry, I seem to be learning that I know even less about PETSc than I thought.

Reply:
Sorry - typo. `petsc does not have an interface to netcdf, parallel-netcdf`

Reply:
* gmp     - nixed
* jpeg
* libpng   (configure checks for zlib as dependencies)
* giflib
* mpfr	   (configure checks for gmp as dependencies)
* netcdf  - nixed
* pnetcdf - nixed
* moab
* eigen   - nixed
* random123
* exodusii (configure checks for hdf5, netcdf, pnetcdf as dependencies)
* mstk    - nixed
* cgns     (configure checks for hdf5 as dependencies)
* memkind
* muparser - nixed (don't know why we have this)
* p4est    (configure checks for zlib as dependencies)
* saws
* libyaml
* zstd	   - nixed (don't know why we have this)

Reply:
> i.e just because `libmpi.so.12` is linked in does not imply the above test code is dependent on MPI

Yes, I understand that. I was trusting the configuration mechanism to use what it was looking for, which was apparently, a mistake.

Reply:
Basically configure is not rejecting mpicc when user specifies it [even-though its not required in this specific case]. And it trusts the user need it for other reasons.. [and checks if things work with it or not]

Reply:
`zstd` can be dropped here. I've been using it in a branch with a student, but that feature hasn't landed yet.

Reply:
So, the missing piece for me was not understanding there was a distinction between packages presented in `configure --help` that interfaced with PETSc and their dependencies. It would be nice if that could be made more clear in the `configure --help` output so that people who are not intimately familiar with PETSc, but just trying to package it for their users, would have more guidance.

The question that I have at this point is should I pursue adding the options from the list that @balay curated to the spack package or just drop this PR?

Reply:
The information is in the configure files; see `self.useddirectly = 0` for those that are only indirect dependencies.
```console
$ git grep useddirecly
```
Yes, someone could make that visible in the `./configure --help` output. Preferably not more lines; the output is long enough already.